src/platforms/common/stm32/swo_manchester_dma.c

```diff
@@ -0,0 +1,570 @@
+	// this is used only by the advanced recovery 
+	timer_ic_set_input(SWO_TIM, TIM_IC1, TIM_IC_IN_TI1);
+	timer_ic_set_input(SWO_TIM, TIM_IC2, TIM_IC_IN_TI1);
+	timer_ic_set_polarity(SWO_TIM, TIM_IC1, TIM_IC_RISING);
+	timer_ic_set_polarity(SWO_TIM, TIM_IC2, TIM_IC_FALLING);
+
+	timer_ic_enable(SWO_TIM, TIM_IC1);
+	timer_ic_enable(SWO_TIM, TIM_IC2);
+#endif
+
+	// interrupt fires twice pre timer cycle (CH3 and UPDATE)
+	// also enable DMA from CH4
+	timer_enable_irq(SWO_TIM, TIM_DIER_UIE | TIM_DIER_CC3IE | TIM_DIER_CC4DE);
+
+	// configure DMA to read edge times into a circular buffer
+	dma_channel_reset(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+
+	dma_set_read_from_peripheral(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+	dma_set_peripheral_size(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, DMA_CCR_PSIZE_16BIT);
+	dma_set_memory_size(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, DMA_CCR_MSIZE_16BIT);
+	dma_set_priority(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, DMA_CCR_PL_HIGH);
+
+	dma_set_peripheral_address(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, (uint32_t)&TIM_CCR4(SWO_TIM));
+	dma_set_memory_address(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, (uint32_t)swo_dma);
+	dma_set_number_of_data(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN, SWO_DMA_EDGE_SAMPLES);
+	dma_enable_circular_mode(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+	dma_enable_memory_increment_mode(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+	dma_enable_channel(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+
+	// enable two DMA interrupts per buffer cycle
+	dma_enable_half_transfer_interrupt(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+	dma_enable_transfer_complete_interrupt(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+
+	// enable DMA interrupt
+	nvic_set_priority(SWO_DMA_EDGE_IRQ, IRQ_PRI_SWO_DMA);
+	nvic_enable_irq(SWO_DMA_EDGE_IRQ);
+
+	// enable timer interrupt
+	nvic_set_priority(SWO_TIM_IRQ, IRQ_PRI_SWO_TIM);
+	nvic_enable_irq(SWO_TIM_IRQ);
+
+	// extra interrupt used for outbound data processing, triggered via NVIC
+	// runs at USB priority to avoid preempting with the regular drain callback
+	nvic_set_priority(SWO_DMA_SW_IRQ, IRQ_PRI_USB);
+	nvic_enable_irq(SWO_DMA_SW_IRQ);
+
+	// start the engine
+	timer_enable_counter(SWO_TIM);
+}
+
+/*
+ * Stops the SWO capture
+ */
+void swo_manchester_deinit(void)
+{
+	timer_disable_counter(SWO_TIM);
+	dma_disable_channel(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+	timer_slave_set_mode(SWO_TIM, TIM_SMCR_SMS_OFF);
+
+	// we can leave the rest of the peripheral configuration alone, just
+	// make sure the restarts in a known state
+	swo_s.s = ST_IDLE;
+	swo_s.rx = 0;
+}
+
+/*
+ * Dedicated handler for processing and outputting trace data
+ * Note the swo_send_buffer is also a callback for the USB SWO_ENDPOINT
+ * that gets called from the main USB ISR, so this one must have the
+ * same priority to avoid accidental preemption
+ */
+void SWO_DMA_SW_ISR(void)
+{
+	// proceed only if there is a chance to send more data
+	if ((*USB_EP_REG(SWO_ENDPOINT) & USB_EP_TX_STAT) != USB_EP_TX_STAT_VALID)
+	{
+		swo_send_buffer(usbdev, SWO_ENDPOINT);
+	}
```

Comment:
The braces are not necessary, please drop them. Please also run this file through `clang-format`.

Reply:
clang-format is now really applied :) I incorrectly presumed `make clang-format` does it since it changed a ton of files, but it goes only two levels deep.